store/postgres/src/chain_head_listener.rs

```diff
@@ -65,56 +77,65 @@ impl ChainHeadUpdateListener {
-                        serde_json::from_value(notification.payload.clone()).unwrap_or_else(|_| {
-                            panic!(
-                                "invalid chain head update received from database: {:?}",
-                                notification.payload
-                            )
-                        });
-
-                    // Observe the latest chain_head_number for each network in order to monitor
-                    // block ingestion
-                    metrics.set_chain_head_number(
-                        &update.network_name,
-                        *&update.head_block_number as i64,
-                    );
-                    futures03::future::ok(Some(update))
-                })
-                .try_for_each(move |update| {
-                    futures03::future::ready(update_sender.send(update).map_err(|_| ()))
-                }),
-        );
+        graph::spawn(async move {
+            while let Some(notification) = receiver.recv().await {
+                // Create ChainHeadUpdate from JSON
+                let update: ChainHeadUpdate =
+                    match serde_json::from_value(notification.payload.clone()) {
+                        Ok(update) => update,
+                        Err(e) => {
+                            crit!(
+                                logger,
+                                "invalid chain head update received from database";
+                                "payload" => format!("{:?}", notification.payload),
+                                "error" => e.to_string()
+                            );
+                            continue;
+                        }
+                    };
+
+                // Observe the latest chain head for each network to monitor block ingestion
+                metrics
+                    .set_chain_head_number(&update.network_name, *&update.head_block_number as i64);
+
+                // If there are subscriptions for this network, notify them.
+                if let Some(watcher) = watchers.read().await.get(&update.network_name) {
+                    watcher.send()
+                }
+            }
+        });
 
         // We're ready, start listening to chain head updates
         listener.start();
     }
-}
 
-impl ChainHeadUpdateListenerTrait for ChainHeadUpdateListener {
-    fn subscribe(&self, network_name: String) -> ChainHeadUpdateStream {
-        let f = move |update: ChainHeadUpdate| {
-            if update.network_name == network_name {
-                futures03::future::ready(Some(()))
+    pub async fn subscribe(&self, network_name: String) -> ChainHeadUpdateStream {
+        let update_receiver = {
+            let existing = {
+                let watchers = self.watchers.read().await;
+                watchers.get(&network_name).map(|w| w.receiver.clone())
+            };
+
+            if let Some(watcher) = existing {
+                watcher
```

Comment:
I tripped over this for about 15s: might be better to call this variable `receiver`